// @arc-fusion/cli/templates/docker-compose.verify.yml.js
'use strict'

const path = require('path')

const verifyDockerfile = require('./verify.Dockerfile')

const { writeFile } = require('../bin/utils/promises')

module.exports = async ({ PROJECT_ROOT, REPO_NAME }) => {
  await writeFile(
    path.join(PROJECT_ROOT, '.fusion', 'verify.Dockerfile'),
    await verifyDockerfile({ PROJECT_ROOT })
  )

  return {
    version: '3',
    services: {
      verify: {
        build: {
          context: '..',
          dockerfile: './.fusion/verify.Dockerfile',
          args: {
            FUSION_RELEASE: `\${FUSION_RELEASE:-latest}`
          }
        },
        command: 'build:prod',
        container_name: 'fusion-verify',
        env_file: [
          '../.env'
        ],
        environment: {
          AWS_XRAY_ENABLED: 'false',
          NODE_ENV: 'production',
          ENVIRONMENT: 'localhost',
          CONTEXT_PATH: `\${CONTEXT_PATH:-pf}`,
          // because we import ALL variables from .env, ignore PORT
          PORT: '8080'
        }
      }
    }
  }
}
