// @arc-fusion/cli/templates/2.0/docker-compose.zip.yml.js
'use strict'

const path = require('path')

const zipDockerfile = require('./zip.Dockerfile')

const { writeFile } = require('../../bin/utils/promises')

module.exports = async ({ ADMIN_RELEASE, FUSION_ROOT, PROJECT_ROOT, REPO_NAME }) => {
  await writeFile(
    path.join(PROJECT_ROOT, '.fusion', 'zip.Dockerfile'),
    zipDockerfile()
  )

  return {
    version: '3',
    services: {
      zip: {
        build: {
          context: '..',
          dockerfile: './.fusion/zip.Dockerfile',
          args: {
            FUSION_RELEASE: `\${FUSION_RELEASE:-latest}`
          }
        },
        container_name: 'fusion-zip',
        env_file: [
          '../.env'
        ],
        environment: {
          AWS_XRAY_ENABLED: 'false',
          NODE_ENV: 'production',
          ENVIRONMENT: 'localhost',
          CONTEXT_PATH: `\${CONTEXT_PATH:-pf}`,
          // because we import ALL variables from .env, ignore PORT
          PORT: '8080'
        },
        network_mode: 'none',
        volumes: [
          '../dist:/opt/zip/dist:rw'
        ]
      }
    }
  }
}
